sweep/fee_bumper.go

```diff
@@ -0,0 +1,142 @@
+
+const (
+	// TxPublished is sent when the broadcast attempt is finished.
+	TxPublished BumpEvent = iota
+
+	// TxFailed is sent when the broadcast attempt fails.
+	TxFailed
+
+	// TxReplaced is sent when the original tx is replaced by a new one.
+	TxReplaced
+
+	// TxConfirmed is sent when the tx is confirmed.
+	TxConfirmed
+
+	// sentinalEvent is used to check if an event is unknown.
+	sentinalEvent
+)
+
+// String returns a human-readable string for the event.
+func (e BumpEvent) String() string {
+	switch e {
+	case TxPublished:
+		return "Published"
+	case TxFailed:
+		return "Failed"
+	case TxReplaced:
+		return "Replaced"
+	case TxConfirmed:
+		return "Confirmed"
+	default:
+		return "Unknown"
+	}
+}
+
+// Unknown returns true if the event is unknown.
+func (e BumpEvent) Unknown() bool {
+	return e >= sentinalEvent
+}
+
+// BumpRequest is used by the caller to give the Bumper the necessary info to
+// create and manage potential fee bumps for a set of inputs.
+type BumpRequest struct {
+	// Budget givens the total amount that can be used as fees by these
+	// inputs.
+	Budget btcutil.Amount
+
+	// Inputs is the set of inputs to sweep.
+	Inputs []input.Input
+
+	// DeadlineHeight is the block height at which the tx should be
+	// confirmed.
+	DeadlineHeight int32
+
+	// DeliveryAddress is the script to send the change output to.
+	DeliveryAddress []byte
+
+	// MaxFeeRate is the maximum fee rate that can be used for fee bumping.
+	MaxFeeRate chainfee.SatPerKWeight
+}
+
+// BumpResult is used by the Bumper to send updates about the tx being
+// broadcast.
+type BumpResult struct {
+	// Event is the type of event that the result is for.
+	Event BumpEvent
+
+	// Tx is the tx being broadcast.
+	Tx *wire.MsgTx
+
+	// ReplacedTx is the old, replaced tx if a fee bump is attempted.
+	ReplacedTx *wire.MsgTx
+
+	// FeeRate is the fee rate used for the new tx.
+	FeeRate chainfee.SatPerKWeight
+
+	// Fee is the fee paid by the new tx.
+	Fee btcutil.Amount
+
+	// Err is the error that occurred during the broadcast.
+	Err error
+}
+
+// Validate validates the BumpResult so it's safe to use.
+func (b *BumpResult) Validate() error {
+	// Every result must have a tx.
+	if b.Tx == nil {
+		return fmt.Errorf("%w: nil tx", ErrInvalidBumpResult)
+	}
+
+	// Every result must have a known event.
+	if b.Event.Unknown() {
+		return fmt.Errorf("%w: unknown event", ErrInvalidBumpResult)
+	}
+
+	// If it's a replacing event, it must have a replaced tx.
+	if b.Event == TxReplaced && b.ReplacedTx == nil {
+		return fmt.Errorf("%w: nil replacing tx", ErrInvalidBumpResult)
+	}
+
+	// If it's a failed event, it must have an error.
+	if b.Event == TxFailed && b.Err == nil {
+		return fmt.Errorf("%w: nil error", ErrInvalidBumpResult)
+	}
+
+	// If it's a confirmed event, it must have a fee rate and fee.
+	if b.Event == TxConfirmed && (b.FeeRate == 0 || b.Fee == 0) {
+		return fmt.Errorf("%w: missing fee rate or fee",
+			ErrInvalidBumpResult)
+	}
+
+	return nil
+}
```

Comment:
I also have questions about this `Validate` method as to where to actually call it. However, I don't think being pointers is the cause - we can replace them with concrete types, then you still have this issue because by default Go just gives these fields empty values. In that sense, using pointers is better in the sense we can easily check against `nil`, rather than figuring out the empty values and compare. Even with `fn.Option`, we still have this issue, because we need to figure out what to do when it's none. I think essentially you can't get away with validations, it has to happen somewhere.

Reply:
Yeah I think validations either happen in a method like this, or as part of a constructor. In this case, we don't have a constructor, so the validate method is used. Alternatively, we could make a constructor that returns an error to hide/meld in the validation.

Reply:
It's not about pointers, per-say. It is about not modeling it as a sum type. We can leave this as is. I am not going to die on this hill right now, but I want to point out that in the case where you have a type that can be a number of different cases where each one of those cases has some guaranteed properties about it, you can use a sum type. That said, sum types in go are a bit inconvenient to model. As is we can't really do better than an interface with a sealed method and then type-switching on it.